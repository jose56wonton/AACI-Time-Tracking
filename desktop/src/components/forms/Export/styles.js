const styles = theme => ({
  hero: {
    height: '100vh',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center'
  },
  textField: { width: '100%' },
  heroContent: {
    width: '100%',
    maxWidth: '500px'
  },
  spaceAround: {
    marginLeft: '8px'
  },
  row: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  error: {
    color: theme.palette.error.main
  },
  title: {
    paddingBottom: '16px !important'
  }
});

export default styles;
