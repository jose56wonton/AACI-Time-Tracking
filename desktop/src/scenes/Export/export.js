import React, { Component } from 'react';

import ExportContainer from 'containers/Export/exportContainer';

class Export extends Component {
  render() {
    return <ExportContainer />;
  }
}

export default Export;
