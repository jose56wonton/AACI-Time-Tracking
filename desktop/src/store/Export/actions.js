import moment from 'moment';

import { exportActionTypes } from 'constants/ActionTypes';
import { employeeActions, projectActions, projectTaskActions, taskActions, shiftActions } from 'store/actions';

import { employeeSelectors, shiftSelectors, projectTaskSelectors, projectSelectors } from 'store/selectors';
import { minutesToString } from 'helpers/time';
import { store } from 'index';

import * as IPCConstants from 'constants/ipc';
import { snackActions } from 'store/actions';
import * as status from 'constants/status';

const electron = window.require('electron');
const ipcRenderer = electron.ipcRenderer;

export const exportToExcel = (exportCategory, start, fileLocation) => {
  return async dispatch => {
    dispatch({ type: exportActionTypes.EXPORT_EXCEL_REQUEST });
    try {

      //console.log(exportCategory,new moment(start).format('YYYY-MM-DD HH:mm:ss'),new moment(end).toString(),fileLocation);

      const startMoment = new moment(start).format('YYYY-MM-DD HH:mm:ss');
      const endMoment = new moment(start).add(7,'days').format('YYYY-MM-DD HH:mm:ss');

      await dispatch(getData(startMoment, endMoment));
      const exportData = formatData(startMoment, endMoment);

      ipcRenderer.sendSync(IPCConstants.CREATE_EXPORT, { fileLocation, data: exportData });
      await dispatch(snackActions.openSnack(status.SUCCESS, 'Export Success!'));
      return dispatch({ type: exportActionTypes.EXPORT_EXCEL_SUCCESS });
    } catch (e) {
      console.log(e);
      dispatch(snackActions.openSnack(status.FAILURE, 'Export failed!'));
      return dispatch({
        type: exportActionTypes.EXPORT_EXCEL_FAILURE,
        payload: e
      });
    }
  };
};

export const getData = (startTime, endTime) => {
  // TODO: different actions based on the export category
  return async dispatch => {
    try {
      await Promise.all([
        dispatch(employeeActions.getEmployees()), dispatch(projectActions.getProjects()), dispatch(projectTaskActions.getProjectTask()), dispatch(taskActions.getTasks()), dispatch(shiftActions.getShiftsInRange(startTime, endTime))
      ]);
    } catch (e) {
      console.log(e);
    }
  };
};

const formatData = (startTime, endTime) => {
  // TODO: different formatting routines for the export category
  // array of employees
  const employees = employeeSelectors.getAllEmployees(store.getState())
    // sort so that employees get added in order
    .sort((a,b)=> a.lastName + a.firstName > b.lastName + b.firstName);
  // array of shifts w/ embedded activities
  const shifts = shiftSelectors
    .getShiftsInRange(store.getState(), { startTime, endTime })
    .sort((a,b)=> a.clockInDate > b.clockInDate);
  // object of project tasks indexed by id with task and project attached
  const projectTasks = projectTaskSelectors.getAllProjectTasksObjects(store.getState());
  const projects = projectSelectors.getAllProjectObjects(store.getState());

  let exportData = [];
  employees.forEach(employee => {
    const detailData = [], summaryData = [];
    const shiftsOfEmployees = shifts.filter(shift => {
      return employee.id === shift.employeeId;
    });


    const individualProjectTotals = {};
    const allProjectTotals = { total: 0, reg: 0, ot: 0 };

    // Summary 
    let totalTimeForWeek = 0;
    shiftsOfEmployees.forEach(shift => {
      shift.activities.forEach(activity => {

        // Adding individual activity times 
        let currentProjectId = projectTasks[activity.projectTaskId].project.id;

        // if project hasn't been added to total 
        if (!individualProjectTotals[currentProjectId]) {
          individualProjectTotals[currentProjectId] = { total: 0, reg: 0, ot: 0 };
        }

        if (totalTimeForWeek >= 2400) {
          allProjectTotals.total += activity.length;
          allProjectTotals.ot += activity.length;
          individualProjectTotals[currentProjectId].total += activity.length;
          individualProjectTotals[currentProjectId].ot += activity.length;

        } else if (totalTimeForWeek + activity.length >= 2400) {

          allProjectTotals.total += activity.length;
          allProjectTotals.reg += (2400 - totalTimeForWeek);
          allProjectTotals.ot += (totalTimeForWeek + activity.length - 2400);
          individualProjectTotals[currentProjectId].total += activity.length;
          individualProjectTotals[currentProjectId].reg += (2400 - totalTimeForWeek);
          individualProjectTotals[currentProjectId].ot += (totalTimeForWeek + activity.length - 2400);

        } else {

          allProjectTotals.total += activity.length;
          allProjectTotals.reg += activity.length;
          individualProjectTotals[currentProjectId].total += activity.length;
          individualProjectTotals[currentProjectId].reg += activity.length;

        }
        totalTimeForWeek += activity.length;
      });
    });
    // add all summary rows
    for (let key in individualProjectTotals) {
      summaryData.push([projects[key].name, '', '', '', '', minutesToString(individualProjectTotals[key].reg), minutesToString(individualProjectTotals[key].ot), minutesToString(individualProjectTotals[key].total)]);
    }
    // add the total summary row
    summaryData.push(['Total', '', '', '', '', minutesToString(allProjectTotals.reg), minutesToString(allProjectTotals.ot), minutesToString(allProjectTotals.total)]);


    // Details 
    totalTimeForWeek = 0;
    shiftsOfEmployees.forEach(shift => {
      shift.activities.forEach((activity, i) => {
        let overtimeActivityLength, regularActivityLength;

        if (totalTimeForWeek >= 2400) {
          regularActivityLength = activity.length;
        } else if (totalTimeForWeek + activity.length >= 2400) {
          overtimeActivityLength = (totalTimeForWeek + activity.length - 2400);
          regularActivityLength = (2400 - totalTimeForWeek);
        } else {
          overtimeActivityLength = activity.length;
        }

        totalTimeForWeek += activity.length;
        if (i === 0) {
          detailData.push([
            moment(shift.clockInDate).format('MM/DD/YYYY'), moment(shift.clockInDate).format('h:mm a'), moment(shift.clockOutDate).format('h:mm a'), minutesToString(shift.lunch), projectTasks[activity.projectTaskId].project.name, projectTasks[activity.projectTaskId].task.name, minutesToString(regularActivityLength), minutesToString(overtimeActivityLength)]);
        } else {
          detailData.push(['','','','',projectTasks[activity.projectTaskId].project.name, projectTasks[activity.projectTaskId].task.name, minutesToString(regularActivityLength), minutesToString(overtimeActivityLength)]);
        }
      });
    });


    // Style
    const spacerRows = Object.keys(individualProjectTotals).length;
    const sheetStyles = {
      1: [1],
      2: [6, spacerRows + 11],
      3: [2, 3, 7, spacerRows + 8, spacerRows + 12],
      spacerRows
    };

    exportData.push({
      key: `${employee.firstName} ${employee.lastName}`,
      header: [
        ['AACI - Time Sheet'], [`Employee: ${employee.firstName} ${employee.lastName}`], [`Period: ${moment(startTime).format('YYYY/MM/DD')} - ${moment(endTime).format('YYYY/MM/DD')}`]
      ],
      summary: [
        [''], [''], ['Summary','', '', '', ''], ['Project', '', '', '', '', 'Reg', 'OT', 'Total'], ...summaryData
      ],
      details: [
        [''], [''], ['Details'], ['Date', 'Clock In', 'Clock Out', 'Lunch', 'Project', 'Task', 'Reg', 'OT'], ...detailData
      ],
      sheetStyles
    });
  });
  return exportData;
};
