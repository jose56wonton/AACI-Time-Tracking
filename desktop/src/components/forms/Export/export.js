import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import { Grid, Button, Typography } from '@material-ui/core';
import { Field, Form } from 'formik';

import styles from './styles';

import TextField from 'components/inputs/TextField';
import File from 'components/inputs/File';
import Select from 'components/inputs/Select';

import * as exportConstants from 'constants/export';

class ExportForm extends Component {
  render() {
    const { classes, cancel, isSubmitting, errors,values } = this.props;    
  
    return (
      <div className={classes.hero}>
        <Form className={classes.heroContent}>
          <Grid container spacing={24}>
            <Grid item xs={12} className={classes.title}>
              <Typography variant="h3">Export</Typography>
            </Grid>
            <Grid item xs={12} className={classes.row}>
              <Field
                name="exportCategory"
                component={Select}
                items={exportConstants.exportCategory}
                fullWidth
                margin="none"
                label="Export By"
                helper="normal"
              />
              <Field
                name="start"
                component={TextField}              
                margin="none"
                label="From"
                type="date"
                helper="normal"
              />
            </Grid>
            <Grid item xs={12}>
              <Field
                name="fileLocation"
                component={File}
                label="File"
                fullWidth
                margin="none"
                type="file"
              />
            </Grid>
            <Grid item xs={12} className={classes.row}>
              <Typography variant="body1" className={classes.error}>{errors[Object.keys(errors)[0]]}</Typography>
              <div>
                <Button
                  type="submit"
                  color="primary"
                  disabled={isSubmitting || Object.keys(errors).length !== 0 || values.fileLocation.length === 0 }
                  variant="contained"
                >
                  Export
                </Button>
                <Button
                  onClick={cancel}
                  color="secondary"
                  variant="text"
                  className={classes.spaceAround}
                >
                  Cancel
                </Button>
              </div>
            </Grid>
          </Grid>
        </Form>
      </div>
    );
  }
}

ExportForm.propTypes = {
  values: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired,
  cancel: PropTypes.func,
  isSubmitting: PropTypes.bool.isRequired,
  errors: PropTypes.object
};
export default withStyles(styles)(ExportForm);
