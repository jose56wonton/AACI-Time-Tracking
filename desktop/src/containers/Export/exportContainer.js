import React, { Component } from 'react';

import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { Formik } from 'formik';
import moment from 'moment';

import Export from 'components/forms/Export';

import { exportValidation } from 'constants/formValidation';
import { exportActions } from 'store/actions';

class ExportContainer extends Component {
  cancel = () => {
    this.props.history.goBack();
  };
  render() {
    return (
      <Formik
        initialValues={{
          exportCategory: 0,
          start: moment('2018-09-22').format('YYYY-MM-DD'),
          end: moment('2018-09-29').format('YYYY-MM-DD'),
          timeLength: 0,
          timeLengthType: 0,
          fileLocation: ''
        }}
        onSubmit={values => {
          const { exportToExcel, history } = this.props;
          const { exportCategory, start, end, fileLocation } = values;
          exportToExcel(exportCategory, start, end, fileLocation)
            .then(() => history.push('/'));
        }}
        validationSchema={exportValidation}
        render={formikProps => {
          return <Export cancel={this.cancel} {...formikProps} />;
        }}
      />
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    exportToExcel: (exportCategory, start, end, fileLocation ) => {
      return dispatch(
        exportActions.exportToExcel(exportCategory, start, end, fileLocation),
      );
    }
  };
};

export default withRouter(
  connect(
    null,
    mapDispatchToProps,
  )(ExportContainer),
);
