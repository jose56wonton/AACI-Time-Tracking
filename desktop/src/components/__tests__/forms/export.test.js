import React from 'react';
import { shallow } from 'enzyme';

import { Export } from 'components/forms/Export/export';
import ExportHOC from 'components/forms/Export';

const props =  {  
  classes: {},
  isSubmitting: true,
  values: {
    fileLocation: '123'
  },
  cancel: jest.fn(),
  errors: {}
};

const setup = overRides => {  
  return shallow(<Export {...props} {...overRides}/>);    
};

const setupHOC = overRides => {
  return shallow(<ExportHOC {...props} {...overRides}/>);
};

describe.skip('Export Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should render correctly', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();    
  });
  it('should render correctly withStyles', () => {
    const wrapper = setupHOC();
    expect(wrapper).toMatchSnapshot();
  });
});