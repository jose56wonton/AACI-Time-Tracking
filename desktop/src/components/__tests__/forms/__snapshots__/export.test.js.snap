// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Component should render correctly 1`] = `
<div>
  <Form>
    <WithStyles(Grid)
      container={true}
      spacing={24}
    >
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(Typography)
          variant="h3"
        >
          Export
        </WithStyles(Typography)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <C
          component={[Function]}
          fullWidth={true}
          helper="normal"
          items={
            Array [
              Object {
                "id": 0,
                "name": "Payroll - Weekly",
              },
            ]
          }
          label="Export By"
          margin="none"
          name="exportCategory"
        />
        <C
          component={[Function]}
          helper="normal"
          label="From"
          margin="none"
          name="start"
          type="date"
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <C
          component={[Function]}
          fullWidth={true}
          label="File"
          margin="none"
          name="fileLocation"
          type="file"
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(Typography)
          variant="body1"
        />
        <div>
          <WithStyles(Button)
            color="primary"
            disabled={true}
            type="submit"
            variant="contained"
          >
            Export
          </WithStyles(Button)>
          <WithStyles(Button)
            color="secondary"
            onClick={[Function]}
            variant="text"
          >
            Cancel
          </WithStyles(Button)>
        </div>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </Form>
</div>
`;

exports[`Export Component should render correctly withStyles 1`] = `
<Export
  cancel={[Function]}
  classes={
    Object {
      "error": "Export-error-6",
      "hero": "Export-hero-1",
      "heroContent": "Export-heroContent-3",
      "row": "Export-row-5",
      "spaceAround": "Export-spaceAround-4",
      "textField": "Export-textField-2",
      "title": "Export-title-7",
    }
  }
  errors={Object {}}
  isSubmitting={true}
  values={
    Object {
      "fileLocation": "123",
    }
  }
/>
`;
